keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -543,56 +686,132 @@ type ReplicaLimit struct {
+`Equivalent` but may be set to `Exact`. See "Use Case: Multiple policy
+definitions for different versions of CRD" for an explanation of why we need
+`MatchPolicy`.
 
-- RBAC, priority & fairness, and admission webhook configurations all use
-ordered lists of rules, and first match allows/matches.
-- Audit configuration also uses an ordered list of rules, but can make a match
-exclude auditing... this approach could address the "ability to skip specific
-resource types" mentioned above.
+xref:
+https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
 
-TODO: What exact matching mechanisms will be supported? Right now this KEP
-shows the use of the same match rules that admission webhooks use.
 
-TODO: Should policy definitions also have match criteria? Match criteria at the
-policy definition level could be used constrain what resources the policy may
-apply to across the entire cluster (like the above "kind constraint" but more
-generally). This would allow policy authors to constrain use of the policy to
-the resources it was designed for.
+Alternative considered: Define match criteria with CEL expressions
 
-TODO: Should "standalone policies" be supported? I.e. should it be possible to
-declare a policy definition that has no `spec.config` but has it's own match
-criteria? Right now a policy requires a minimum of three resources to do
-anything (policy definition, configuration CRD, at least one policy
-configuration).
+Pros:
 
-`MatchPolicy` will implicitly be `Equivalent`, and will not be configurable,
-but it will be possible to opt-out of the behavior by inspecting the
-GVK of the admitted resource in the CEL expression.
+- CEL expressions can be used to declare far more sophisticated match criteria.
 
-xref:
-https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#matching-requests-rules
+Cons:
+
+- CEL expressions can evalaute to an error while API types only evaluate to
+  "matches" / "does not match" result.
+- We intend to allow policy configuration resources declare match criteria,
+  making it difficult to pre-compile or type-check CEL expressions.
+
+##### Exclude matching
+
+Benefits:
+
+- Exclude rules make it easier for policy definition authors to do things like
+  exclude kube-system or other special purpose namespaces from policies that apply
+  to all "normal" namespaces.
+- Default rules make it easier to apply principal of least privilege to
+  policies. I.e. ensure that newly created resources default to a least
+  privilege state until configured otherwise.
+
+Exclude rules can be modeled after the approach exstablished in Audit where
+level=None on a match rule can be used to exlude any rules beneath it. E.g.
+`effect: Exclude`.
+
+If we want this, it is essential that we introduce it early, the exclude inverts
+the semantics of a match, adding it in a later release is _very_ incompatible.
+
+#### Singleton policies
```

Comment:
Adding this complicates reporting/debugging/analyzing the policy space somewhat. To the extent that this is purely a UX benefit (I don't think it allows expression of anything otherwise unexpressable), the downsides should also be considered.

Reply:
This seems manageable. If we report most violations using a {policy definition, policy configuration} identifier pair (presumably names since they are both cluster scope in phase 1), then for singletons we could just use {policy definition, policy definition} as the identifier pair. It keeps things consistent and makes tracing from violation info back to the resource that produced it straight-forward.